File: Week2/22.py
from functools import lru_cache
class Solution:
    @lru_cache(None)
    def generateParenthesis(self, n: int) -> List[str]:
        if n==1:
            return ["()"]
        else:
            res=set()
            temp=self.generateParenthesis(n-1)
            for s in temp:
                for i in range(len(s)):
                    ss=s[:i]+'()'+s[i:]
                    res.add(ss)
            return list(res)

File: Week2/22.js
var map={0:[""]}
var generateParenthesis = function(n) {
    if (n in map){
        return map[n];
    }
    if (n==0){
        return [""];
    }
    var res=[]
    for(let k = 1; k<=n; k++){
        var a=generateParenthesis(k-1)
        var b=generateParenthesis(n-k)
        for (const i of a){
            for (const j of b){
                res.push(`(${i})${j}`)
            }
        }
    }
    map[n]=res
    return res
};